web/federate.go

```diff
@@ -317,3 +310,66 @@ Loop:
 		}
 	}
 }
+
+// makeNativeHistogram creates a dto.Histogram representing a native histogram.
+// Use only for standard exponential schemal.
+func makeNativeHistogram(h *histogram.FloatHistogram) *dto.Histogram {
+	result := &dto.Histogram{
+		SampleCountFloat: proto.Float64(h.Count),
+		SampleSum:        proto.Float64(h.Sum),
+		Schema:           proto.Int32(h.Schema),
+		ZeroThreshold:    proto.Float64(h.ZeroThreshold),
+		ZeroCountFloat:   proto.Float64(h.ZeroCount),
+		NegativeCount:    h.NegativeBuckets,
+		PositiveCount:    h.PositiveBuckets,
+	}
+	if len(h.PositiveSpans) > 0 {
+		result.PositiveSpan = make([]*dto.BucketSpan, len(h.PositiveSpans))
+		for i, sp := range h.PositiveSpans {
+			result.PositiveSpan[i] = &dto.BucketSpan{
+				Offset: proto.Int32(sp.Offset),
+				Length: proto.Uint32(sp.Length),
+			}
+		}
+	}
+	if len(h.NegativeSpans) > 0 {
+		result.NegativeSpan = make([]*dto.BucketSpan, len(h.NegativeSpans))
+		for i, sp := range h.NegativeSpans {
+			result.NegativeSpan[i] = &dto.BucketSpan{
+				Offset: proto.Int32(sp.Offset),
+				Length: proto.Uint32(sp.Length),
+			}
+		}
+	}
+	return result
+}
+
+// makeClassicHistogram creates a dto.Histogram representing a classic
+// histogram. Use only for NHCB (schema -53).
+func makeClassicHistogram(h *histogram.FloatHistogram) *dto.Histogram {
+	result := &dto.Histogram{
+		SampleCountFloat: proto.Float64(h.Count),
+		SampleSum:        proto.Float64(h.Sum),
+	}
+	result.Bucket = make([]*dto.Bucket, len(h.CustomValues))
+	var (
+		cumulativeCount float64
+		bucketIter      = h.PositiveBucketIterator()
+		bucketAvailable = bucketIter.Next()
+	)
+	for i, le := range h.CustomValues {
+		for bucketAvailable && int(bucketIter.At().Index) < i {
+			bucketAvailable = bucketIter.Next()
+		}
+		if bucketAvailable && int(bucketIter.At().Index) == i {
+			cumulativeCount += bucketIter.At().Count
+		}
+		result.Bucket[i] = &dto.Bucket{
+			UpperBound:           proto.Float64(le),
+			CumulativeCountFloat: proto.Float64(cumulativeCount),
+		}
+	}
+	// Note that we do not add the +Inf bucket explicitly. In the protobuf
+	// exposition format, it is optional.
```

Comment:
I don't think this doc string is correct.

It is true that in protobuf +Inf is [optional](https://github.com/prometheus/prometheus/blob/55a4782eb7ad95bee116fc4255c94d5cdfc1a82f/model/textparse/protobufparse.go#L696), however we don't restrict NHCB to protobuf exposition, so you'd also have to add that in PromText and OpenMetrics +Inf is also optional and it is the exposition that adds +Inf bucket.

- OpenMetrics: https://github.com/prometheus/common/blob/0df7b916149edfcc9e6099b8352e7d311fa221e4/expfmt/openmetrics_create.go#L350
- Prometheus text: https://github.com/prometheus/common/blob/0df7b916149edfcc9e6099b8352e7d311fa221e4/expfmt/text_create.go#L247